R/UWerr.R

```diff
@@ -240,98 +239,110 @@ uwerrderived <- function(f, data, nrep, S=1.5, pl=FALSE, ...) {
-        flag=0
+      W=W+1
+    }
+    
+    if(flag) {
+      warning("Windowing condition failed!")
+      Wopt[[i]] <- Wmax[[i]]
+    }
+    
+    CFbbopt <- GammaFbb[[i]][1] + 2*sum(GammaFbb[[i]][2:(Wopt[[i]]+1)])
+    if(CFbbopt <= 0) {
+      ##stop("Gamma pathological, estimated error^2 <0\n")
+      warning("Gamma pathological, estimated error^2 <0\n")
+    }
+    GammaFbb[[i]] <- GammaFbb[[i]] + CFbbopt/N # bias in Gamma corrected
+    CFbbopt <- GammaFbb[[i]][1] + 2*sum(GammaFbb[[i]][2:(Wopt[[i]]+1)]) #refined estimate
+    sigmaF <- sqrt(abs(CFbbopt)/N) #error of F
+    rho <- GammaFbb[[i]]/GammaFbb[[i]][1]
+    tauintFbb <- cumsum(rho)-0.5 #normalised autocorrelation
+    
+    ## bias cancellation for the mean value
+    
+    if(R >= 2) {
+      bF <- (Fb[i]-Fbb[i])/(R-1);
+      Fbb <- Fbb[i] - bF;
+      if(abs(bF) > sigmaF/4) {
+        warning("a %.1f sigma bias of the mean has been cancelled", bF/sigmaF)
       }
+      
+      Fbr[i] = Fbr[i] - bF*N/nrep;
+      Fb[i]  = Fb[i]  - bF*R;
     }
-    W=W+1
-  }
-
-  if(flag) {
-    warning("Windowing condition failed!")
-    Wopt <- Wmax
-  }
-  
-  CFbbopt <- GammaFbb[1] + 2*sum(GammaFbb[2:(Wopt+1)])
-  if(CFbbopt <= 0) {
-    ##stop("Gamma pathological, estimated error^2 <0\n")
-    warning("Gamma pathological, estimated error^2 <0\n")
-  }
-  GammaFbb <- GammaFbb+CFbbopt/N # bias in Gamma corrected
-  CFbbopt <- GammaFbb[1] + 2*sum(GammaFbb[2:(Wopt+1)]) #refined estimate
-  sigmaF <- sqrt(abs(CFbbopt)/N) #error of F
-  rho <- GammaFbb/GammaFbb[1]
-  tauintFbb <- cumsum(rho)-0.5 #normalised autocorrelation
+    
+    res$value[i] <- Fbb[i]
+    res$dvalue[i] <- sigmaF
+    res$ddvalue[i] <- sigmaF*sqrt((Wopt[[i]]+0.5)/N)
+    res$tauint[i]  <- tauintFbb[(Wopt[[i]]+1)]
+    res$dtauint[i] = res$tauint[i]*2*sqrt((Wopt[[i]]-res$tauint[i]+0.5)/N)
 
-  ## bias cancellation for the mean value
+    tauintofW[[i]] <- tauintFbb[1:(Wmax[[i]]+1)]
+    dtauintofW[[i]] <- tauintFbb[1:(Wmax[[i]]+1)]*sqrt(c(0:Wmax[[i]])/N)*2 
 
-  if(R >= 2) {
-    bF <- (Fb-Fbb)/(R-1);
-    Fbb <- Fbb - bF;
-    if(abs(bF) > sigmaF/4) {
-      warning("a %.1f sigma bias of the mean has been cancelled", bF/sigmaF)
+    ## Q value for replica distribution if R>=2
+    if(R>1) {
+      chisqr <- sum((Fbr[i]-Fb[i])^2*nrep)/CFbbopt
+      res$Qval[i] <- 1-pgamma(chisqr/2, (R-1)/2)
     }
-    
-    Fbr = Fbr - bF*N/nrep;
-    Fb  = Fb  - bF*R;
   }
-
-  value <- Fbb
-  dvalue <- sigmaF
-  ddvalue <- sigmaF*sqrt((Wopt+0.5)/N)
-  tauint  <- tauintFbb[(Wopt+1)]
-  dtauint = tauint*2*sqrt((Wopt-tauint+0.5)/N)
-  dtauintofW <- tauintFbb[1:(Wmax+1)]*sqrt(c(0:Wmax)/N)*2 
   
-  # Q value for replica distribution if R>=2
-  if(R>1) {
-    chisqr <- sum((Fbr-Fb)^2*nrep)/CFbbopt
-    Qval <- 1-pgamma(chisqr/2, (R-1)/2)
-  }
-  else {
-    Qval <- NULL
-  }
+  res <- list(res=res, Wopt=Wopt, Wmax=Wmax,
+              tauintofW=tauintofW, dtauintofW=dtauintofW,
+              S=S, fgrad=fgrad, datamean=abb,
```

Comment:
would it not be more consistent to either have all determined quantities (value, dvalue, ddvalue, tauint, Wopt) which have the dimension ```length(Fbb)``` to either be individual elements of the result list or, conversely, all be bundled into the ```res``` sublist?

Reply:
ping?

Reply:
@urbach 
Just to clarify what I would like to do. Currently the return values of uwerrprimary and uwerrderived have different structures. In particular, the latter contains a ```res``` list element which needs to be accessed. I would like to modify the return value of uwerrderived to contain ```value, dvalue, ddvalue, tauint, ddtauint, Qval``` as vectors (and not part of ```res```. This way, uwerrprimary and uwerrderived have consistent return values, up to the fact that for the latter, these are vectors rather than single numbers.

Reply:
See https://github.com/urbach/hadron/pull/52